############################################
##
## Purpose of script: compare transcriptome assemblies from two different RNA-seq data
##
## Author: Lucia Lorenzi
##
## Date Created: 2024-09-12
##
## Email: lucialorenzi90@gmail.com
##
#  Notes ---------------------------
##
##
##
#  Setup ---------------------------

options(scipen = 999)
# Load packages---------------------------
# load up the packages we will need:  (uncomment as required)
require(tidyverse)
require(ggvenn)
library(trastools)
library(UpSetR)

setwd("~/Rprojects/StemLinc_analyses/")

# Inputs :
# ***Modify this part to include your desired inputs***-----

# 1. tracking file of the combined transcriptome from both samples
tracking_path="outputs/gffcompare/T_cellvsMacro.tracking"

# 2. annotated tracking files generated by "Transcriptome_characterization_report.Rmd"
tr1_path="outputs/transcriptome_characterization/T_cell.combined/T_cell.combined_annotated_tracking.tsv"
tr2_path="outputs/transcriptome_characterization/Macro.combined/Macro.combined_annotated_tracking.tsv"
# 3. sample names
sample_names=c("T_cell","Macro")

# 4. reference annotation file
ref_annot_path="data/references/merged_refs_annotation/annotated_tracking_file.updated_gene_names.txt"


# End of edits ----
# Load data---------------------------
library(ggsci)
nejm_pal=ggsci::pal_nejm()(8)
tracking=read.table(tracking_path)
tr1=read.table(tr1_path,header = T)
tr2=read.table(tr2_path,header = T)
ref_annot=read.table(ref_annot_path,header = T)
samp1=sample_names[1]
samp2=sample_names[2]
# functions ----

plot_pair_venn <- function(tra,
                           samp1_var,
                           samp2_var,
                           feat_var,
                           filter_cond=rep(T,nrow(tra)), # defaults the entire data
                           sampnames=sample_names,
                           title=""){

  # Capture the column names as symbols
  samp1_var <- ensym(samp1_var)
  samp2_var <- ensym(samp2_var)
  feat_var <- ensym(feat_var)

  tra <- tra %>% filter(filter_cond)
  if (is.logical(tra %>% pull(!!samp1_var))) {
    vennlist <- list(
      tra %>% filter(!!samp1_var) %>% pull(!!feat_var),
      tra %>% filter(!!samp2_var) %>% pull(!!feat_var)
    )
  } else {
    vennlist <- list(
      tra %>% filter(!!samp1_var != "-") %>% pull(!!feat_var),
      tra %>% filter(!!samp2_var != "-") %>% pull(!!feat_var)
    )
  }
  names(vennlist)=sampnames
  g=ggvenn(
    vennlist,
    fill_color = c("#0072B5FF","#FFDC91FF"),
    stroke_size = 0.5, set_name_size = 4
  ) +ggtitle(title)
  print(g)
  #return(list(vennlist,g))
} # makes venn diagram of two sets of transcripts

plot_three_venn <- function(traGL,
                            samp1_var,
                            samp2_var,
                            samp3_var,
                            feat_var,
                            filter_cond=rep(T,nrow(traGL)), # defaults the entire data
                            sampnames=c(sample_names,"Ref"),
                            title=""){
  traGL <- traGL %>% filter(filter_cond)
  if(is.logical(traGL%>%pull({{samp1_var}}))){
    vennlist <- list(traGL %>% filter({{samp1_var}}) %>% pull({{feat_var}}) ,
                     traGL %>% filter({{samp2_var}}) %>% pull({{feat_var}}),
                     traGL %>% filter({{samp3_var}}) %>% pull({{feat_var}}))
  }else{
    vennlist <- list(traGL %>% filter({{samp1_var}}!="-") %>% pull({{feat_var}}) ,
                     traGL %>% filter({{samp2_var}}!="-") %>% pull({{feat_var}}),
                     traGL %>% filter({{samp3_var}}!="-") %>% pull({{feat_var}}))
  }

  names(vennlist)=sampnames

  g=ggvenn(
    vennlist,
    fill_color = c("#0072B5FF","#FFDC91FF","#20854EFF"),
    stroke_size = 0.5, set_name_size = 4
  ) +ggtitle(title)
  print(g)
  #return(list(vennlist,g))
} # makes venn diagram of three sets of transcripts

get_max_per_group_by_arrange <- function(tracking,
                                         group_var,
                                         arr_var,
                                         summ_var={{arr_var}},
                                         descending = T,
                                         outname="max_val"){

  tracking <- tracking %>% group_by({{group_var}})

  if(descending){
    tracking <- tracking %>% arrange(desc({{arr_var}}))
  } else tracking <- tracking %>% arrange({{arr_var}})

  tracking <- tracking %>% summarise(!!outname:= {{summ_var}}[1])

  return(tracking)
}

summarize_GL <- function(tracking,
                         ol_cc=overlapping_class_codes,
                         sampnames=sample_names,
                         ordered_classcodes=c("=",
                                              "j",
                                              "k",
                                              "c",
                                              "m",
                                              "n",
                                              "e",
                                              "o",
                                              "p",
                                              "s",
                                              "x",
                                              "i",
                                              "y",
                                              "u",
                                              "r",
                                              ".")){

  samp1=sampnames[1]
  samp2=sampnames[2]

  # first assign gene name from ref, if present, XLOC if not
  tracking <- tracking %>% mutate(gene_name = ifelse(V4%in%ol_cc,
                                                     ifelse(!is.na(ref_gene_name_1),
                                                            ref_gene_name_1,
                                                            ref_gene_name_2),
                                                     V2))
  # group by gene_name summarise each sample occurrence

  tracking_GL <- tracking %>% group_by(gene_name) %>% summarise(!!samp1:=any(q1_gene!="-"),
                                                                !!samp2:=any(q2_gene!="-"))

  # get maximum values per gene for variables of interest
  variables= c("mean_tpm_1","mean_tpm_2","Nsamps_1","Nsamps_2","Nexons_1","Nexons_2")

  max_vals <- map(variables, function(var) {
    onam=paste0("max_",var)
    var=sym(var)
    get_max_per_group_by_arrange(tracking,
                                 group_var = gene_name,
                                 arr_var = !!var,
                                 outname = onam)
  })

  tracking_GL <- left_join(tracking_GL,Reduce(merge,max_vals))

  # best classcode per sample

  c2 <- tracking %>% mutate(!!samp1:=V5!="-",!!samp2:=V6!="-") %>%
    pivot_longer(cols = c(!!samp1,!!samp2),names_to = "sample",values_to = "is") %>%
    filter(is) %>% group_by(gene_name,sample) %>% arrange(match(V4,ordered_classcodes)) %>%
    summarise(best_cc=V4[1])
  c3 <- c2 %>% pivot_wider(names_from = sample,values_from = best_cc)
  c3[is.na(c3)] = "NA" # try to preserve the sample order
  c3 <- c3 %>% select(gene_name,!!samp1,!!samp2)
  colnames(c3)[2:3]=paste0(colnames(c3)[2:3],"_cc")

  tracking_GL <- left_join(tracking_GL,c3)

  return(tracking_GL)
}

annotate_tracking_with_2_datasets <- function(tracking, tr1, tr2, cols=c(5:ncol(tracking))){
  # things to add
  #     - ref transcript biotype
  #     - Number of samples in which each transcript is assembled in each dataset (from each sample's tracking)
  #     - Number of exons of each transcript (from tracking)
  #     - expression of each transcript in each dataset (from each sample's tracking)

  trs=list(tr1,tr2)
  trs=lapply(trs,function(t){
    t=t%>%dplyr::select(V1,ref_gene_name,ref_biotype,mean_tpm,Nexons,Nsamps)
    return(t)
  })
  trs=lapply(seq_along(trs),function(i){
    t=trs[[i]]
    colnames(t)[2:ncol(t)]=paste0(colnames(t)[2:ncol(t)],"_",i)
    return(t)
  })



  tracking <- trastools::split_samples_info(tracking, cols , remove = F)
  tracking <- left_join(tracking,trs[[1]],by=c(q1_transcript="V1"))
  tracking <- left_join(tracking,trs[[2]],by=c(q2_transcript="V1"))

  return(tracking)
}



#filtering by any desired condition (e.g classcode, exonic type, etc)


# Annotate tracking file ----
tracking <- annotate_tracking_with_2_datasets(tracking = tracking,tr1,tr2)

# Transcript level comparison ----

plot_pair_venn(tracking,V5,V6,V1,

               title = "Transcript level exact match")

# per classcode
for (cc in c("=","c","e","i","j","k","m","n","o","p","u","x")) {
  plot_pair_venn(tracking,V5,V6,V1,
                 filter_cond = tracking$V4==cc,sampnames = sample_names,
                 title = paste0("Transcript level overlap of ",cc," transcripts"))
}

plot_pair_venn(tracking,V5,V6,V1,
               filter_cond = tracking$V4%in%overlapping_class_codes,
               title = "Transcripts that overlap Ref exact match")
plot_pair_venn(tracking,V5,V6,V1,
               filter_cond = !tracking$V4%in%overlapping_class_codes,
               title = "Transcripts that do not overlap Ref exact match")

# Gene level comparison ----


## Gene level summarization ----
tracking_GL <- summarize_GL(tracking,sampnames = sample_names)


# add biotype of reference
tracking_GL$biotype <- ref_annot$simplified_gene_biotype[match(tracking_GL$gene_name,
                                                               ref_annot$gene_name)]

## Gene level plots ----

### Heatmaps classcodes ----
myplots <- htmltools::tagList()
plot_heatmap <- function(cc_dat,sampnames=sample_names,
                         title=""){
  samp1=sampnames[1]
  samp2=sampnames[2]
  cc_dat <- cc_dat %>%
    mutate(text = paste0(samp1,": ", Var1, "\n", samp2,": ", Var2,
                         "\n", "Value: ",Freq, "\n", "Frac:", round(Freq/sum(Freq),2)))
  p=ggplot(cc_dat, aes(Var1, Var2, fill= Freq,text=text)) +
    geom_tile() + ggtitle(title) + xlab(samp1) +
    ylab(samp2)

 # p=plotly::as_widget(plotly::ggplotly(p, tooltip="text"))
  return(p)
}



cc_dat=as.data.frame(table(as.data.frame(tracking_GL)[,colnames(tracking_GL)==paste0(samp1,"_cc")],
                           as.data.frame(tracking_GL)[,colnames(tracking_GL)==paste0(samp2,"_cc")]))


p=plot_heatmap(cc_dat = cc_dat, title = "All assembled genes")
myplots[[1]]=plotly::as_widget(plotly::ggplotly(p, tooltip="text"))

i=2
for (biot in unique(tracking_GL$biotype)) {
  tra=tracking_GL%>%filter(biotype%in%biot)
  cc_dat=as.data.frame(table(as.data.frame(tra)[,colnames(tra)==paste0(samp1,"_cc")],
                             as.data.frame(tra)[,colnames(tra)==paste0(samp2,"_cc")]))


  p=plot_heatmap(cc_dat,title = paste0("Assembled genes in ",biot,
                                     " biotype"))
  myplots[[i]] <- plotly::as_widget(plotly::ggplotly(p, tooltip="text"))
  i=i+1
}
myplots
### Pairwise Venns ----
plot_pair_venn(tracking_GL, !!sym(samp1), !!sym(samp2),
               gene_name, title = "Gene level overlap")
plot_pair_venn(tracking_GL,!!sym(samp1),!!sym(samp2),gene_name,
               filter_cond = !is.na(tracking_GL$biotype),
               title = "Gene level overlap - annotated genes")

plot_pair_venn(tracking_GL,!!sym(samp1),!!sym(samp2),gene_name,
               filter_cond = is.na(tracking_GL$biotype),
               title = "Gene level overlap - potential novel genes")


for (biot in c("protein_coding","lncRNA","pseudogene")) {
  plot_pair_venn(tracking_GL,
                 !!sym(samp1),
                 !!sym(samp2),
                   gene_name,
                   filter_cond = !is.na(tracking_GL$biotype)&
                   tracking_GL$biotype==biot,
                   title = paste0("Gene level overlap - ",biot))

}

### Venns including ref ----
tracking_GL_with_ref <- full_join(ref_annot %>%
                                    mutate(biotype=simplified_gene_biotype,
                                           Ref=TRUE) %>%
                                    dplyr::select(gene_name,biotype,Ref),
                             tracking_GL %>%
                               dplyr::select(gene_name,!!sym(samp1),!!sym(samp2),biotype)
                                  )

tracking_GL_with_ref[is.na(tracking_GL_with_ref)] <- FALSE

plot_three_venn(tracking_GL_with_ref,!!sym(samp1),!!sym(samp2),
                Ref,gene_name,title = "Gene level overlap")

for (biot in unique(tracking_GL_with_ref$biotype)) {
  p=plot_three_venn(tracking_GL_with_ref,!!sym(samp1),
                    !!sym(samp2),Ref,gene_name,
                  title = paste0("Gene level overlap - ",
                                 biot," and potential novel"),
                  filter_cond = tracking_GL_with_ref$biotype%in%
                    c(biot,"FALSE"))
  print(p)
}

### Barplot of number of samples in each dataset ----
# per the top X most abundant groups in the heatmap

cc_dat$combined_cc=paste(cc_dat$Var1,cc_dat$Var2,sep = ";")
X=15
top_classes=cc_dat %>% arrange(-Freq) %>%slice(1:X)
top_classes

tracking_GL$combined_cc=paste(as.data.frame(tracking_GL)[,colnames(tracking_GL)==paste0(samp1,"_cc")],
                              as.data.frame(tracking_GL)[,colnames(tracking_GL)==paste0(samp2,"_cc")],sep = ";")

tra <- tracking_GL %>% filter(combined_cc%in%top_classes$combined_cc)
tra <- tra%>%dplyr::select(combined_cc, max_Nsamps_1,max_Nsamps_2) %>%
  pivot_longer(cols = 2:3,names_to = "sample",values_to = "maxNsamps") %>%
  mutate(sample = recode(sample, "max_Nsamps_1" = samp1, "max_Nsamps_2" = samp2))

tra <- tra[!is.na(tra$maxNsamps),]

ggplot(tra,aes(x=sample,fill=as.factor(maxNsamps))) +
  geom_bar() + facet_wrap(~combined_cc)
ggplot(tra,aes(x=sample,fill=as.factor(maxNsamps))) +
  geom_bar(position = "fill") + facet_wrap(~combined_cc)


### UpSet plots ----
#### Intronic genes and number of samples ----
# intronic genes vs number of samples,

intronic_genes <- tracking_GL %>% filter(combined_cc%in%c("i;i","i;NA","NA;i"))
Samp1_intronic=lapply(1:3,function(i)intronic_genes%>%
                           filter(max_Nsamps_1==i) %>% pull(gene_name))
Samp2_intronic=lapply(1:3,function(i)intronic_genes%>%
                        filter(max_Nsamps_2==i) %>% pull(gene_name))
intronic_genes_list <- c(Samp1_intronic,Samp2_intronic)
names(intronic_genes_list)=c(paste0(samp1,"_intronic_maxSamps=",1:3),
                             paste0(samp2,"_intronic_maxSamps=",1:3))

upset(fromList(intronic_genes_list), order.by = "freq",nsets = 6)
grid.text("Intronic genes and max number of assembled samples",x = 0.65, y=0.95, gp=gpar(fontsize=14))

#### Exonic type ----
exonic_type_list <- list(Samp1_mono=tracking_GL%>%filter(max_Nexons_1==1) %>% pull(gene_name),
                         Samp1_multi=tracking_GL%>%filter(max_Nexons_1>1) %>% pull(gene_name),
                         Samp2_mono=tracking_GL%>%filter(max_Nexons_2==1) %>% pull(gene_name),
                         Samp2_multi=tracking_GL%>%filter(max_Nexons_2>1) %>% pull(gene_name))

names(exonic_type_list) <- c(paste0(samp1,c("_mono","_multi")),
                                  paste0(samp2,c("_mono","_multi")))
sets <- fromList(exonic_type_list)
sets$gene_name=unique(unlist(exonic_type_list,use.names = F))
tracking_GL$overlapRef=!is.na(tracking_GL$biotype)
sets$overlapRef=tracking_GL$overlapRef[match(sets$gene_name,
                                             tracking_GL$gene_name)]
upset(sets,
      query.legend = "bottom", nsets = 6, number.angles = 30, point.size = 3.5, line.size = 2,
      queries = list(

        list(
          query = elements,
          params = list("overlapRef",T),
          color =  "#20854EFF",
          active = T,
          query.name = "Overlaps Reference"
        )
      )
)
grid.text("Exonic type and overlap Ref",x = 0.65, y=0.95, gp=gpar(fontsize=14))

upset(sets %>% filter(!overlapRef),
      query.legend = "bottom", nsets = 6, number.angles = 30, point.size = 3.5, line.size = 2,

)

grid.text("Exonic type of potential novel genes",x = 0.65, y=0.95, gp=gpar(fontsize=14))

## Expression plots ----
tracking_GL <- tracking_GL %>% mutate(gene_class=ifelse(overlapRef,biotype,"potNovel"),
                                      in_sample=ifelse(!!sym(samp1)&!!sym(samp2),"both",
                                                       ifelse(!!sym(samp1),samp1,samp2)))
expression_dat <- tracking_GL%>%filter(gene_class%in%c("potNovel","protein_coding",
                                                       "lncRNA","TEC","pseudogene"))

expression_dat <- pivot_longer(expression_dat,
                               cols = c("max_mean_tpm_1","max_mean_tpm_2"),
                               names_to = "sample", values_to = "TPM")

expression_dat <- expression_dat%>% mutate(sample = recode(sample,max_mean_tpm_1=samp1,
                                                           max_mean_tpm_2=samp2))

expression_dat <- expression_dat%>%filter(!is.na(TPM))
ggplot(expression_dat, aes(x=in_sample,fill=sample,y=TPM)) +geom_boxplot()+
  scale_y_log10() + facet_wrap(~gene_class) + theme_bw()

ggplot(expression_dat, aes(x=in_sample)) +geom_bar()+
  facet_wrap(~gene_class) + theme_bw() + ylab("# genes")

# add max number of samples
expression_dat <- expression_dat %>% mutate(maxNsamps=ifelse(sample==samp1,max_Nsamps_1,
                                                             max_Nsamps_2))

ggplot(expression_dat, aes(x=interaction(in_sample,maxNsamps))) +geom_bar()+
  facet_wrap(~gene_class) + theme_bw() + ylab("# genes") + theme(axis.text.x = element_text(angle = 45,hjust = 1))


ggplot(expression_dat, aes(x=interaction(in_sample,maxNsamps),fill=sample,y=TPM)) +geom_boxplot()+
  scale_y_log10() + facet_wrap(~gene_class) + theme_bw() + theme(axis.text.x = element_text(angle = 45,hjust = 1))

